### PR TITLE
Fix gather_all_tensors

### PR DESCRIPTION
**Pull request by maximsch2**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
gather_all_tensors doesn't work correctly when tensors are different shape, fix that

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by maximsch2** [comment]
@ananthsub, I think it can be great to bring this over to `LightningModule.all_gather` as well to make it easier to use.

**Review by Borda — APPROVED** [review]
lgtm

File: tests/bases/test_ddp.py [hunk 1]
```diff
@@ -56,8 +57,37 @@ def _test_ddp_sum_cat(rank, worldsize):
...
+    setup_ddp(rank, worldsize)
+    tensor = torch.ones(rank)
+    result = gather_all_tensors(tensor)
+    assert len(result) == worldsize
+    for idx in range(worldsize):
+        assert len(result[idx]) == idx
+        assert (result[idx] == torch.ones_like(result[idx])).all()
+
+
+def _test_ddp_gather_uneven_tensors2(rank, worldsize):
```
Comment (SkafteNicki) [inline tests/bases/test_ddp.py:70]:
```suggestion
def _test_ddp_gather_uneven_tensors_multidim(rank, worldsize):
```

File: tests/bases/test_ddp.py [hunk 2]
```diff
@@ -56,8 +57,37 @@ def _test_ddp_sum_cat(rank, worldsize):
...
+
 @pytest.mark.skipif(sys.platform == "win32", reason="DDP not available on windows")
-@pytest.mark.parametrize("process", [_test_ddp_cat, _test_ddp_sum, _test_ddp_sum_cat])
+@pytest.mark.parametrize(
+    "process", [
+        _test_ddp_cat,
+        _test_ddp_sum,
+        _test_ddp_sum_cat,
+        _test_ddp_gather_uneven_tensors,
+        _test_ddp_gather_uneven_tensors2,
```
Comment (SkafteNicki) [inline tests/bases/test_ddp.py:88]:
```suggestion
        _test_ddp_gather_uneven_tensors_multidim,
```

File: torchmetrics/utilities/distributed.py [hunk 3]
```diff
@@ -88,6 +89,12 @@ def class_reduce(num: Tensor, denom: Tensor, weights: Tensor, class_reduction: s
     )
 
 
+def _simple_gather_all_tensors(result: Tensor, group: Any, world_size: int):
```
Comment (SkafteNicki) [inline torchmetrics/utilities/distributed.py:92]:
```suggestion
def _simple_gather_all_tensors(result: Tensor, group: Any, world_size: int) -> List[Tensor]:
    """ Function for gathering a tensor with same shape across processes """
```

File: torchmetrics/utilities/distributed.py [hunk 4]
```diff
@@ -88,6 +89,12 @@ def class_reduce(num: Tensor, denom: Tensor, weights: Tensor, class_reduction: s
...
+    gathered_result = [torch.zeros_like(result) for _ in range(world_size)]
+    torch.distributed.all_gather(gathered_result, result, group)
+    return gathered_result
+
+
 def gather_all_tensors(result: Union[Tensor], group: Optional[Any] = None):
```
Comment (SkafteNicki) [inline torchmetrics/utilities/distributed.py:98]:
```suggestion
def gather_all_tensors(result: Tensor, group: Optional[Any] = None) -> List[Tensor]:
```

File: torchmetrics/utilities/distributed.py [hunk 5]
```diff
@@ -108,11 +115,37 @@ def gather_all_tensors(result: Union[Tensor], group: Optional[Any] = None):
...
+    # 1. Gather sizes of all tensors
+    local_size = torch.tensor(result.shape, device=result.device)
+    local_sizes = [torch.zeros_like(local_size) for _ in range(world_size)]
+    torch.distributed.all_gather(local_sizes, local_size, group=group)
+    max_size = torch.stack(local_sizes).max(dim=0).values
+    all_sizes_equal = True
+    for size in local_sizes:
+        if not (size == max_size).all():
+            all_sizes_equal = False
+            break
```
Comment (SkafteNicki) [inline torchmetrics/utilities/distributed.py:133]:
```suggestion
    all_sizes_equal = all(all(ls == max_size) for ls in local_sizes)
```

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by ananthsub** [comment]
@tchaton : @maximsch2  pointed out we should bring this to Lightning to replace the existing all_gather_tensors utility here: https://github.com/PyTorchLightning/pytorch-lightning/blob/597b309f2e3b0f5fb618f4326e8700df399917fb/pytorch_lightning/utilities/distributed.py#L86-L113

File: torchmetrics/utilities/distributed.py [hunk 6]
```diff
@@ -88,7 +89,13 @@ def class_reduce(num: Tensor, denom: Tensor, weights: Tensor, class_reduction: s
...
+def _simple_gather_all_tensors(result: Tensor, group: Any, world_size: int) -> List[Tensor]:
+    gathered_result = [torch.zeros_like(result) for _ in range(world_size)]
+    torch.distributed.all_gather(gathered_result, result, group)
+    return gathered_result
+
+
+def gather_all_tensors(result: Tensor, group: Optional[Any] = None) -> List[Tensor]:
     """
     Function to gather all tensors from several ddp processes onto a list that
     is broadcasted to all processes
```
Comment (SkafteNicki) [inline torchmetrics/utilities/distributed.py:101]:
```suggestion
    is broadcasted to all processes. Works on tensors that have the same number
    of dimensions, but where each dimension may differ. In this case tensors are
    padded, gathered and then trimmed to secure equal workload for all processes
```

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Pull request merged**